# Week 2

1. Q&A
2. Data frames
    - Creating them
    - Indexes and column names
    - Retrieving and setting
    - Methods and attributes we'll want to use
    - `.loc` and mask indexes
3. Creating data frames from files
    - CSV (comma-separated values / character-separated values)
    - Excel
    - JSON
    - Retrieving data from the Internet


In [3]:
import pandas as pd   # this means: find pandas.py, load it into memory, but refer to it with the "pd" variable
import numpy as np    
from pandas import Series, DataFrame  # this means: find pandas.py, load it into memory, and assign Series and DataFrame

In [4]:
type(pd)

module

# Data frames

If a series is a 1D data structure, with values (that have a dtype) and an index, then a data frame is a 2D data structure. It has rows (the equivalent to a series index) and columns (which also a type of index, but along the tops of the columns).

Each column in a data frame is actually a series.

In [5]:
# create a data frame, using a list of lists

df = DataFrame([[10, 20, 30],
                [40, 50, 60], 
                [70, 80, 90],
                [100, 110, 120]])

df

,0,1,2
0,10,20,30
1,40,50,60
2,70,80,90
3,100,110,120


In [6]:
df.index #  give me the index

RangeIndex(start=0, stop=4, step=1)

In [7]:
df.columns  # give me the columns

RangeIndex(start=0, stop=3, step=1)

In [9]:
# what if I want to set the index or the column names?
# we saw last week that I can pass a list of values for the index
# that continues to be true! And I can do that for the columns, also


df = DataFrame([[10, 20, 30],
                [40, 50, 60], 
                [70, 80, 90],
                [100, 110, 120]],
              index=list('abcd'),
               columns=list('xyz'))

df

,x,y,z
a,10,20,30
b,40,50,60
c,70,80,90
d,100,110,120


In [10]:
df.index

Index(['a', 'b', 'c', 'd'], dtype='str')

In [11]:
df.columns

Index(['x', 'y', 'z'], dtype='str')

In [12]:
# how can I retrieve a row?
# the answer is -- just as we did values in a series
# why? Because the index is per row, just as it was with a series!
# the best way to retrieve? We can use .loc and .iloc

df.loc['b']  # get row with the index b

x    40
y    50
z    60
Name: b, dtype: int64

In [13]:
df.loc[['b', 'd']]  # fancy indexing, asking for more than one row

,x,y,z
b,40,50,60
d,100,110,120


In [14]:
df.iloc[2]  # get the row at position 2

x    70
y    80
z    90
Name: c, dtype: int64

In [15]:
df.iloc[[1, 3]]

,x,y,z
b,40,50,60
d,100,110,120


In [17]:
# remember slices?

df.loc['b':'d']  # slices with .loc are up to and INCLUDING

,x,y,z
b,40,50,60
c,70,80,90
d,100,110,120


In [18]:
df.iloc[1:3]  # slices with .iloc are the traditional up to and NOT INCLUDING

,x,y,z
b,40,50,60
c,70,80,90


In [19]:
s = Series([10, 20, 30], index=list('abc'))
s

a    10
b    20
c    30
dtype: int64

In [21]:
# how can we retrieve one or more columns?
# answer: with [] 
# that's why I told you last week to use .loc to retrieve from a series index, so you won't get confused here!

df['x']  # get one column

a     10
b     40
c     70
d    100
Name: x, dtype: int64

In [23]:
df[['x', 'z']]   # two columns, aka a data frame

,x,z
a,10,30
b,40,60
c,70,90
d,100,120


In [ ]:
# is there a way to retrieve positionally from the columns, similar to .iloc?
# nope, there isn't.

# remember that the index (rows) can be repeated. Column names can be repeated, but Pandas *really* discourages
# you from doing it, and tries to stop you.



# Exercise: Grocery store

1. Create a data frame describing products in a grocery store. The columns will be `product`, `price`, and `department`. You should have 5 total products. Each will be a new row in the data frame. The index should be the numeric ID for that product. Each product's ID is a 3-digit number.
2. Retrieve the products at positions 0 and 3.
3. Retrieve those same products via their indexes.
4. Retrieve just the `price` column. What is the mean price of items in your store?

In [27]:
df = DataFrame([['apple', 5, 'produce'],
                ['bananas', 3, 'produce' ],
                ['computer', 500, 'electronics' ],
                ['diary', 20, 'books'  ],
                ['elephant', 100, 'pets']],
              index=[123, 124, 345, 456, 789],
              columns='product price department'.split())
df
                

,product,price,department
123,apple,5,produce
124,bananas,3,produce
345,computer,500,electronics
456,diary,20,books
789,elephant,100,pets


In [28]:
df.iloc[[0, 3]]  # fancy indexing with iloc

,product,price,department
123,apple,5,produce
456,diary,20,books


In [29]:
df.loc[[123, 456]]  # fancy indexing with .loc

,product,price,department
123,apple,5,produce
456,diary,20,books


In [30]:
# what if I use strings rather than numbers?

df.loc[['123', '456']]  

KeyError: "None of [Index(['123', '456'], dtype='str')] are in the [index]"

In [33]:
# Retrieve just the price column. What is the mean price of items in your store?

df['price'].mean()

np.float64(125.6)

In [34]:
df['price'].median()

np.float64(20.0)

In [35]:
# can I assign a new column to our data frame?
# yes! All I have to do is assign, and give the value as a list (or series) of the right length

#              broadcasting to get a new series back
# assignment
df['tax'] = df['price'] * 0.1
df

,product,price,department,tax
123,apple,5,produce,0.5
124,bananas,3,produce,0.3
345,computer,500,electronics,50.0
456,diary,20,books,2.0
789,elephant,100,pets,10.0


In [36]:
# what if I assign to a column that already exists?
# I overwrite the existing column, and replace it

df['tax'] = df['price'] * 0.2

In [37]:
df

,product,price,department,tax
123,apple,5,produce,1.0
124,bananas,3,produce,0.6
345,computer,500,electronics,100.0
456,diary,20,books,4.0
789,elephant,100,pets,20.0


In [38]:
# what if I want to rename a column?
# data frames have *TONS* of methods, and the odds are that you can do what you want
# in this case, df.rename, and pass columns= with a dict argument, {from:to}

df.rename(columns={'tax':'sales_tax'})

,product,price,department,sales_tax
123,apple,5,produce,1.0
124,bananas,3,produce,0.6
345,computer,500,electronics,100.0
456,diary,20,books,4.0
789,elephant,100,pets,20.0


In [39]:
df

,product,price,department,tax
123,apple,5,produce,1.0
124,bananas,3,produce,0.6
345,computer,500,electronics,100.0
456,diary,20,books,4.0
789,elephant,100,pets,20.0


# Method chaining

Pandas methods generally assume that you don't want to change the data frame, but that you instead want to get a new data frame back, reflecting the changes you made.

For example, in the case of `rename`, we got back a new data frame that was different from `df`, in that a column was renamed. But it didn't change the column permanently.

If you want to change it permanently, you can (should) assign back to the variable

    df = df.rename(.....)

You can (but SHOULD NOT) pass the `inplace=True` keyword argument to a lot of methods. This does change the data frame in place, and returns `None`.

Why not do this?

1. You won't be able to do method chaining, meaning that you won't be able to invoke a bunch of transformations on your data frame.
2. It's often assumed that using `inplace=True` is faster, or uses less memory, than getting a new data frame back. THIS IS NOT THE CASE.


In [40]:
# if I want to remove a column, I can use df.drop, passing a list to the "columns" keyword argument

df.drop(columns=['tax'])   # this returns a new data frame, without 'tax'

,product,price,department
123,apple,5,produce
124,bananas,3,produce
345,computer,500,electronics
456,diary,20,books
789,elephant,100,pets


In [41]:
df = df.drop(columns=['tax'])  # the right way to do it

In [42]:
df

,product,price,department
123,apple,5,produce
124,bananas,3,produce
345,computer,500,electronics
456,diary,20,books
789,elephant,100,pets


# Retrieving with `.loc` -- the 2-argument version

We've used `.loc` to retrieve an element from a series and a row from a data frame.

It turns out  that we can also invoke `.loc` with two arguments:

- The first is a "row selector"
    - one index value
    - a list of index values ("fancy indexing")
    - boolean index to retrieve only those element where it's `True`
- The second is a "column selector"
    - one column name
    - a list of column names

In [43]:
df

,product,price,department
123,apple,5,produce
124,bananas,3,produce
345,computer,500,electronics
456,diary,20,books
789,elephant,100,pets


In [44]:
df.loc[
    123   # row selector
    ,
    'department'   # column selector
    ]

'produce'

In [45]:
df.loc[
    [123, 789]     # row selector
    ,
    'department'   # column selector
    ]

123    produce
789       pets
Name: department, dtype: str

In [46]:
df.loc[
    [123, 789]     # row selector
    ,
    ['department', 'product']   # column selector
    ]

,department,product
123,produce,apple
789,pets,elephant


# Why `{}` or `[]`?

Pandas uses Python core data structures. In many ways, it's logical, but it's not always obvious:

- In `pd.rename`, we pass a dict, because we have the original names and the new names. So a dict seems like a reasonable way to go.
- In `pd.drop`, we're just providing a list of columns we want to drop. There's no reason to have a dict there.

Generally speaking, in Pandas, wherever you can provide one index/column name as as string, you can provide a list of index/column names as multiple strings.

In [47]:
s

a    10
b    20
c    30
dtype: int64

In [48]:
df

,product,price,department
123,apple,5,produce
124,bananas,3,produce
345,computer,500,electronics
456,diary,20,books
789,elephant,100,pets


# Series vs. data frame

A series is 1D -- it has values and an index. Every value in a series has the same dtype. That can be int64, float32, or string.

A data frame is 2D - it has values in rows and columns. The index describes the rows, and the column names describe the columns. 

Every column in a data frame is really a series. Every column has a dtype.



In [49]:
df

,product,price,department
123,apple,5,produce
124,bananas,3,produce
345,computer,500,electronics
456,diary,20,books
789,elephant,100,pets


# Exercise: Family members

1. Create a data frame with 5 family members (or friends, or make it up). Each row is one person. The columns will be named `name`, `age`, and `shoesize`.
2. What is the mean `shoesize` for all members of your family?
3. What are the `min` and `max` ages for members of your family?
4. Retrieve all of the names and their shoe sizes.
5. Retrieve the age and name of the person in row 0.
6. Retrieve the age and name of the people in rows 1 and 3.

In [52]:
df = DataFrame([['Reuven', 55, 46],
                ['SF', 53, 40],
                ['AMLF', 25, 40],
                ['SBLF', 23, 40],
                ['ADLF', 20, 44]],
              columns=['name', 'age', 'shoesize'])
df                

,name,age,shoesize
0,Reuven,55,46
1,SF,53,40
2,AMLF,25,40
3,SBLF,23,40
4,ADLF,20,44


In [53]:
# What is the mean shoesize for all members of your family?

df['shoesize'].mean()

np.float64(42.0)

In [56]:
# What are the min and max ages for members of your family?

df['age'].describe().loc[['min', 'max']]

min    20.0
max    55.0
Name: age, dtype: float64

In [ ]:
# Retrieve all of the names and their shoe sizes.

df[['name', 'shoesize']]   # retrieving more than one column with a list in []

In [57]:
# Retrieve the age and name of the person in row 0.

# we'll use the two-argument version of .loc

df.loc[
    0               # row selector
    ,
    ['age', 'name'] # column selector
    ]

age         55
name    Reuven
Name: 0, dtype: object

In [58]:
# Retrieve the age and name of the people in rows 1 and 3.

df.loc[
    [1, 3]          # row selector
    ,
    ['age', 'name'] # column selector
    ]

,age,name
1,53,SF
3,23,SBLF


# Next up

1. Useful methods and attributes
2. Boolean/mask indexes with data frames

# Useful methods and attributes

Pandas defines a *ton* of methods for use of series, and many (most?) work on data frames, as well. And of course, data frames have many of their own methods.

Generally speaking, a method that runs on series will also run on a data frame:

- If the series method returns a series, the data frame method will return a data frame
- If the series method returns one value, the data frame method will return a series

In [59]:
df = DataFrame([[10, 20, 30],
                [40, 50, 60], 
                [70, 80, 90],
                [100, 110, 120]],
               index=list('abcd'),
               columns=list('xyz'))
df

,x,y,z
a,10,20,30
b,40,50,60
c,70,80,90
d,100,110,120


In [60]:
df['x'].mean()  # get the mean from column x

np.float64(55.0)

In [61]:
df.mean()   # get the mean from each column

x    55.0
y    65.0
z    75.0
dtype: float64

In [62]:
df['y'].max()  # get the max from column y

np.int64(110)

In [63]:
df.max()   # get the max from each column

x    100
y    110
z    120
dtype: int64

In [64]:
df['z'].describe()   # get the descriptive statistics for column z

count      4.000000
mean      75.000000
std       38.729833
min       30.000000
25%       52.500000
50%       75.000000
75%       97.500000
max      120.000000
Name: z, dtype: float64

In [65]:
df.describe()   # get descriptive statistics for every column

,x,y,z
count,4.000000,4.000000,4.000000
mean,55.000000,65.000000,75.000000
std,38.729833,38.729833,38.729833
min,10.000000,20.000000,30.000000
25%,32.500000,42.500000,52.500000
50%,55.000000,65.000000,75.000000
75%,77.500000,87.500000,97.500000
max,100.000000,110.000000,120.000000


In [66]:
df

,x,y,z
a,10,20,30
b,40,50,60
c,70,80,90
d,100,110,120


In [67]:
# add a new column!

df['t'] = ['hello', 'to', 'everyone', 'today']
df

,x,y,z,t
a,10,20,30,hello
b,40,50,60,to
c,70,80,90,everyone
d,100,110,120,today


In [68]:
# how can I make column t into the index for my data frame?
# answer: the set_index method

df.set_index('t')

,x,y,z
t,,,
hello,10,20,30
to,40,50,60
everyone,70,80,90
today,100,110,120


In [69]:
# invoking set_index does *not* change the data frame! Rather, it
# returns a new data frame with the changed index

df

,x,y,z,t
a,10,20,30,hello
b,40,50,60,to
c,70,80,90,everyone
d,100,110,120,today


# Method chaining

In Python, you can typically have one statement/expression per line. However, if you open `()`, Python sees everything inside of them as being on one line.

This lends itself to writing long, complex Pandas queries across multiple lines, where each row "inherits" the output from the previous one.

In [76]:
(
    df
    .set_index('t') 
    .loc[
        ['hello', 'today'],   # row selector -- these two rows
         ['x', 'z']          # column selector -- these two columns
         ]
    .sum()
)

x    110
z    150
dtype: int64

In [81]:
# if you do set the index permanently (by assigning to a variable), you can also
# invoke reset_index, and turn the index back into a regular column

(
    df
    .set_index('t') 
    .loc[['hello', 'today']]
    .reset_index()
    [['t', 'y']]
)

,t,y
0,hello,20
1,today,110


# Remember `nan`?

`NaN` (or `nan`), is "not a number." This is what data sets use to indicate that they're missing values, or the values are unknown.

On a series:

- `dropna` returns a new series without the orignal's `NaN` values
- `fillna` returns a new series where `NaN` is replaced by whatever you give it

These methods also exist on data frames!

- If you invoke `dropna`, then any *row* containing `NaN` is removed
- If you invoke `fillna`, then any `NaN` value is replaced. If you pass a single value as the replacement, then it's used everywhere. But if you pass a series, then you can actually replace different values in different columns.

In [82]:
df

,x,y,z,t
a,10,20,30,hello
b,40,50,60,to
c,70,80,90,everyone
d,100,110,120,today


In [83]:
df = df.drop(columns=['t'])
df

,x,y,z
a,10,20,30
b,40,50,60
c,70,80,90
d,100,110,120


In [84]:
df.loc['b', 'y'] = np.nan
df.loc['a', 'z'] = np.nan
df.loc['a', 'x'] = np.nan
df.loc['d', 'y'] = np.nan

df


,x,y,z
a,NaN,20.0,NaN
b,40.0,NaN,60.0
c,70.0,80.0,90.0
d,100.0,NaN,120.0


In [85]:
# let's check the dtypes of our columns

df.dtypes  

x    float64
y    float64
z    float64
dtype: object

In [86]:
# what happens if I invoke df.dropna()? Every row with even one NaN is removed

df.dropna()  # this returns a new data frame, based on df...

,x,y,z
c,70.0,80.0,90.0


In [87]:
df

,x,y,z
a,NaN,20.0,NaN
b,40.0,NaN,60.0
c,70.0,80.0,90.0
d,100.0,NaN,120.0


In [88]:
df.fillna(999)

,x,y,z
a,999.0,20.0,999.0
b,40.0,999.0,60.0
c,70.0,80.0,90.0
d,100.0,999.0,120.0


In [89]:
df.mean()  # remember, NaN isn't included in this calculation

x    70.0
y    50.0
z    90.0
dtype: float64

In [90]:
df.fillna(df.mean())   # replace NaN in each column with the mean of the non-NaN values

,x,y,z
a,70.0,20.0,90.0
b,40.0,50.0,60.0
c,70.0,80.0,90.0
d,100.0,50.0,120.0


# Exercise: More with groceries

1. (Re)define the data frame with groceries.
2. Set the index to be the product names.
3. Retrieve three items via the index, and find the min/max prices for them.
4. Set the index to be the department. Find the mean price for items in one department.

In [91]:
df = DataFrame([['apple', 5, 'produce'],
                ['bananas', 3, 'produce' ],
                ['computer', 500, 'electronics' ],
                ['diary', 20, 'books'  ],
                ['elephant', 100, 'pets']],
              index=[123, 124, 345, 456, 789],
              columns='product price department'.split())
df



,product,price,department
123,apple,5,produce
124,bananas,3,produce
345,computer,500,electronics
456,diary,20,books
789,elephant,100,pets


In [94]:
# Set the index to be the product names.

df = df.set_index('product')
df

,price,department
product,,
apple,5,produce
bananas,3,produce
computer,500,electronics
diary,20,books
elephant,100,pets


In [95]:
df

,price,department
product,,
apple,5,produce
bananas,3,produce
computer,500,electronics
diary,20,books
elephant,100,pets


In [100]:
# Retrieve three items via the index, and find the min/max prices for them.

(
    df
    .loc[
        ['apple', 'computer', 'elephant'],  # row selector
        'price'   # column selector
        ]
    .describe()
    .loc[['min', 'max']]
)

min      5.0
max    500.0
Name: price, dtype: float64

In [101]:
# Set the index to be the department. Find the mean price for items in one department.

# Boolean indexes

You might remember from last week that Pandas uses a weird kind of syntax, a "boolean index" (aka a "mask index") to retrieve selected values from a series.

The idea is:

- We invoke, via broadcasting, an operator on a series
- We get back a series of `True` and `False` values
- We apply that boolean series, via `.loc`, to our series of values
- Only those that correspond to `True` are kept

In [103]:
s.loc[[True, False, True]]  # very simple version

a    10
c    30
dtype: int64

In [106]:
s.loc[ s >= 20 ]  # get all values of s that are >= 20

b    20
c    30
dtype: int64

In [107]:
# we can do the same thing with a data frame!
# our row selector (i.e., the first argument to .loc) can be a boolean series!

df

,price,department
product,,
apple,5,produce
bananas,3,produce
computer,500,electronics
diary,20,books
elephant,100,pets


In [108]:
df['price'] > 30

product
apple       False
bananas     False
computer     True
diary       False
elephant     True
Name: price, dtype: bool

In [109]:
# traditional way to select rows in Pandas

df.loc[   df['price'] > 30  ]

,price,department
product,,
computer,500,electronics
elephant,100,pets


In [110]:
# as of Pandas 3.0 (from January), we have a new, better syntax

df.loc[   pd.col('price') > 30  ]

,price,department
product,,
computer,500,electronics
elephant,100,pets


# Exercises

1. Use grocery-store data.
2. Add a new column, `tax`, which is the percentage tax applied to each item. Vary it according to the department.
3. Add a new column, `total_price`, which is the price + tax.
4. Retrieve products where `total_price` > 50.
5. Retrieve products where `department` is `produce`.

In [112]:
df = df.reset_index()
df

,product,price,department
0,apple,5,produce
1,bananas,3,produce
2,computer,500,electronics
3,diary,20,books
4,elephant,100,pets


In [113]:
df['tax'] = [0.1, 0.1, 0.2, 0.15, 0.3]
df

,product,price,department,tax
0,apple,5,produce,0.10
1,bananas,3,produce,0.10
2,computer,500,electronics,0.20
3,diary,20,books,0.15
4,elephant,100,pets,0.30


In [114]:
df['total_price'] = df['price'] * df['tax']
df

,product,price,department,tax,total_price
0,apple,5,produce,0.10,0.5
1,bananas,3,produce,0.10,0.3
2,computer,500,electronics,0.20,100.0
3,diary,20,books,0.15,3.0
4,elephant,100,pets,0.30,30.0


In [116]:
# Retrieve products where total_price > 50.

df.loc[   df['total_price'] > 50  ]

,product,price,department,tax,total_price
2,computer,500,electronics,0.2,100.0


In [117]:
df.loc[   pd.col('total_price') > 50 ]

,product,price,department,tax,total_price
2,computer,500,electronics,0.2,100.0


In [118]:
# Retrieve products where department is produce.

df.loc[   df['department'] == 'produce'  ]

,product,price,department,tax,total_price
0,apple,5,produce,0.1,0.5
1,bananas,3,produce,0.1,0.3


In [119]:
df.loc[   pd.col('department') == 'produce' ]

,product,price,department,tax,total_price
0,apple,5,produce,0.1,0.5
1,bananas,3,produce,0.1,0.3


In [ ]:
# what if I want rows where total_price < 50 *and* department is produce?

(
    df.loc[  pd.col('department') == 'produce'  ]   # this returns a data frame
    df.loc[  pd.col('total_price') < 50   ]         # this uses line 4's data frame
)

# Next up

1. More with boolean indexes and methods
2. Real-world data (with files!)

Download this: https://files.lerner.co.il/data-science-exercise-files.zip

In [120]:
df

,product,price,department,tax,total_price
0,apple,5,produce,0.10,0.5
1,bananas,3,produce,0.10,0.3
2,computer,500,electronics,0.20,100.0
3,diary,20,books,0.15,3.0
4,elephant,100,pets,0.30,30.0


In [124]:
# how would a more complex query look?

# for example: I want the mean price for products whose tax rate is > 12%
# and whose price > 25

(
    df
    .loc[  pd.col('tax') > 0.12 ]
    .loc[  pd.col('price') > 25 ]
    ['price']
    .mean()
)

np.float64(300.0)

In [125]:
# better, use the 2-argument version of .loc

(
    df
    .loc[  pd.col('tax') > 0.12 ]
    .loc[  pd.col('price') > 25, 'price' ]  # 2-argument version
    .mean()
)

np.float64(300.0)

# Real-world data and loading info from files

When I started to work with data, I was sure that there were great, well-thought-out file formats. That is now starting to be the case, but it usually isn't.

The most common file format is CSV (comma-separated values), which is basically a table on disk:

- Every line in the file is a row in the data frame
- Every line broken up (usually with commas) into fields
- Every line has the same number of fields.

The problem is that everything in a CSV file is text. So Pandas has to figure out which columns are integers and floats, and it often can't do that on its own.

# Files

I unzipped all of the files from that zipfile into the same directory as Jupyter. If you're using Jupyter lite, you'll need to drag from your computer's desktop into the left-hand pane in Jupyter lite.

How can we read from a file into a data frame? Pandas provides a *huge* number of methods, each for reading from a different file format into a data frame.

In [126]:
# the methods for reading from a file into a data frame are all pd.read_* methods

df = pd.read_csv('taxi.csv')  # NYC taxi rides in 2016 -- a (very bad) sample of 10,000 rides

In [127]:
# head returns the first 5 rows

df.head()

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,pickup_longitude,pickup_latitude,RateCodeID,store_and_fwd_flag,dropoff_longitude,dropoff_latitude,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
0,2,2015-06-02 11:19:29,2015-06-02 11:47:52,1,1.63,-73.954430,40.764141,1,N,-73.974754,40.754093,2,17.0,0.0,0.5,0.00,0.0,0.3,17.80
1,2,2015-06-02 11:19:30,2015-06-02 11:27:56,1,0.46,-73.971443,40.758942,1,N,-73.978539,40.761909,1,6.5,0.0,0.5,1.00,0.0,0.3,8.30
2,2,2015-06-02 11:19:31,2015-06-02 11:30:30,1,0.87,-73.978111,40.738434,1,N,-73.990273,40.745438,1,8.0,0.0,0.5,2.20,0.0,0.3,11.00
3,2,2015-06-02 11:19:31,2015-06-02 11:39:02,1,2.13,-73.945892,40.773529,1,N,-73.971527,40.760330,1,13.5,0.0,0.5,2.86,0.0,0.3,17.16
4,1,2015-06-02 11:19:32,2015-06-02 11:32:49,1,1.40,-73.979088,40.776772,1,N,-73.982162,40.758999,2,9.5,0.0,0.5,0.00,0.0,0.3,10.30


In [128]:
df.describe()

,VendorID,passenger_count,trip_distance,pickup_longitude,pickup_latitude,RateCodeID,dropoff_longitude,dropoff_latitude,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
count,9999.000000,9999.000000,9999.000000,9999.000000,9999.000000,9999.000000,9999.000000,9999.000000,9999.000000,9999.000000,9999.000000,9999.000000,9999.000000,9999.000000,9999.000000,9999.000000
mean,1.516652,1.659466,3.158511,-73.014956,40.226521,1.045105,-73.054699,40.248644,1.407741,14.415892,0.118212,0.497450,1.818059,0.400433,0.299790,17.552472
std,0.499748,1.333306,4.037516,8.347871,4.599169,0.302132,8.186847,4.510520,0.501911,12.442624,0.214794,0.037667,2.634469,1.665170,0.010816,15.137990
min,1.000000,0.000000,0.000000,-74.186302,0.000000,1.000000,-74.277367,0.000000,1.000000,-7.000000,-0.500000,-0.500000,0.000000,0.000000,-0.300000,-7.800000
25%,1.000000,1.000000,1.000000,-73.990997,40.738556,1.000000,-73.990261,40.738478,1.000000,7.000000,0.000000,0.500000,0.000000,0.000000,0.300000,8.800000
50%,2.000000,1.000000,1.700000,-73.979774,40.755909,1.000000,-73.978256,40.756340,1.000000,10.500000,0.000000,0.500000,1.000000,0.000000,0.300000,12.800000
75%,2.000000,2.000000,3.300000,-73.963001,40.770012,1.000000,-73.961311,40.771044,2.000000,17.000000,0.000000,0.500000,2.460000,0.000000,0.300000,19.800000
max,2.000000,6.000000,64.600000,0.000000,41.064606,5.000000,0.000000,41.137344,4.000000,250.000000,1.000000,0.500000,42.050000,70.000000,0.300000,252.350000


# Exercise: Weird taxi rides

1. Read `taxi.csv` into a data frame. The file contains 9,999 taxi rides in NYC.
2. How many rides when 0 miles (`trip_distance`)? How much, on average, did people pay (`total_amount`) for such rides?
3. How many rides <= 0 dollars? How far (`trip_distance`), on average, did people travel on such trips?
4. How many rides had 0 passengers (`passenger_count`)? Give the mean for `trip_distance` and `total_amount`.

In [129]:
df = pd.read_csv('taxi.csv')

In [130]:
df

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,pickup_longitude,pickup_latitude,RateCodeID,store_and_fwd_flag,dropoff_longitude,dropoff_latitude,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
0,2,2015-06-02 11:19:29,2015-06-02 11:47:52,1,1.63,-73.954430,40.764141,1,N,-73.974754,40.754093,2,17.0,0.0,0.5,0.00,0.0,0.3,17.80
1,2,2015-06-02 11:19:30,2015-06-02 11:27:56,1,0.46,-73.971443,40.758942,1,N,-73.978539,40.761909,1,6.5,0.0,0.5,1.00,0.0,0.3,8.30
2,2,2015-06-02 11:19:31,2015-06-02 11:30:30,1,0.87,-73.978111,40.738434,1,N,-73.990273,40.745438,1,8.0,0.0,0.5,2.20,0.0,0.3,11.00
3,2,2015-06-02 11:19:31,2015-06-02 11:39:02,1,2.13,-73.945892,40.773529,1,N,-73.971527,40.760330,1,13.5,0.0,0.5,2.86,0.0,0.3,17.16
4,1,2015-06-02 11:19:32,2015-06-02 11:32:49,1,1.40,-73.979088,40.776772,1,N,-73.982162,40.758999,2,9.5,0.0,0.5,0.00,0.0,0.3,10.30
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9994,1,2015-06-01 00:12:59,2015-06-01 00:24:18,1,2.70,-73.947792,40.814972,1,N,-73.973358,40.783638,2,11.0,0.5,0.5,0.00,0.0,0.3,12.30
9995,1,2015-06-01 00:12:59,2015-06-01 00:28:16,1,4.50,-74.004066,40.747818,1,N,-73.953758,40.779285,1,16.0,0.5,0.5,3.00,0.0,0.3,20.30
9996,2,2015-06-01 00:13:00,2015-06-01 00:37:25,1,5.59,-73.994377,40.766102,1,N,-73.903206,40.750546,2,21.0,0.5,0.5,0.00,0.0,0.3,22.30
9997,2,2015-06-01 00:13:02,2015-06-01 00:19:10,6,1.54,-73.978302,40.748531,1,N,-73.989166,40.762852,2,6.5,0.5,0.5,0.00,0.0,0.3,7.80


In [131]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9999 entries, 0 to 9998
Data columns (total 19 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   VendorID               9999 non-null   int64  
 1   tpep_pickup_datetime   9999 non-null   str    
 2   tpep_dropoff_datetime  9999 non-null   str    
 3   passenger_count        9999 non-null   int64  
 4   trip_distance          9999 non-null   float64
 5   pickup_longitude       9999 non-null   float64
 6   pickup_latitude        9999 non-null   float64
 7   RateCodeID             9999 non-null   int64  
 8   store_and_fwd_flag     9999 non-null   str    
 9   dropoff_longitude      9999 non-null   float64
 10  dropoff_latitude       9999 non-null   float64
 11  payment_type           9999 non-null   int64  
 12  fare_amount            9999 non-null   float64
 13  extra                  9999 non-null   float64
 14  mta_tax                9999 non-null   float64
 15  tip_amount     

In [135]:
# How many rides went 0 miles (trip_distance)? How much, on average, did people pay (total_amount) for such rides?

(
    df
    .loc[ 
        pd.col('trip_distance') == 0 ,    # row selector
        'total_amount'                    # column selector
        ]
    .mean()
)

np.float64(31.581940298507465)

In [138]:
# How many rides <= 0 dollars? How far (trip_distance), on average, did people travel on such trips?

(
    df
    .loc[
        pd.col('total_amount') <= 0,    # row selector
        'trip_distance'
        ]
    .mean()
)

np.float64(0.6066666666666667)

In [141]:
# How many rides had 0 passengers (passenger_count)? Give the mean for trip_distance and total_amount.

(
    df
    .loc[
        pd.col('passenger_count') == 0,    # row selector
        ['trip_distance', 'total_amount']  # column selector
        ]
    .mean()
    
)

trip_distance     4.60
total_amount     25.57
dtype: float64

# Options for `read_csv`

CSV is a very loose standard. There are many variants on it in the world. For that reason, `read_csv` needs to be flexible enough to work with them. There are *MANY* parameters for `read_csv`, most of which are optional and keyword-only, meaning that you can to pass `name=value` to `read_csv` for them to work.

In [142]:
help(pd.read_csv)

Help on function read_csv in module pandas:

read_csv(
    filepath_or_buffer: FilePath | ReadCsvBuffer[bytes] | ReadCsvBuffer[str],
    *,
    sep: str | None | lib.NoDefault = <no_default>,
    delimiter: str | None | lib.NoDefault = None,
    header: int | Sequence[int] | None | Literal['infer'] = 'infer',
    names: Sequence[Hashable] | None | lib.NoDefault = <no_default>,
    index_col: IndexLabel | Literal[False] | None = None,
    usecols: UsecolsArgType = None,
    dtype: DtypeArg | None = None,
    engine: CSVEngine | None = None,
    converters: Mapping[HashableT, Callable] | None = None,
    true_values: list | None = None,
    false_values: list | None = None,
    skipinitialspace: bool = False,
    skiprows: list[int] | int | Callable[[Hashable], bool] | None = None,
    skipfooter: int = 0,
    nrows: int | None = None,
    na_values: Hashable | Iterable[Hashable] | Mapping[Hashable, Iterable[Hashable]] | None = None,
    keep_default_na: bool = True,
    na_filter: bool 

### Most common `read_csv` options

1. `usecols` -- pass a list of column names you want in your data frame. Anything you don't mention is ignored.
2. `index_col` -- which column from the file should be used as your index?
3. `header` -- on which row of the file are the header names? By default, it's 0, meaning the first line. You can set it to any integer or to `None`, which means: No names are provided.
4. `sep` -- what is the separator between fields? By default, it's `','`, but it's common to have `\t` (tab)

In [144]:
# let's say that Pandas makes a mistake when reading your file, and you have to change 
# the dtype of a column. What do you do?

# you can use the 'astype' method on any series, INCLUDING A COLUMN! 

df['trip_distance'] = df['trip_distance'].astype('float32')

In [145]:
df

,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,pickup_longitude,pickup_latitude,RateCodeID,store_and_fwd_flag,dropoff_longitude,dropoff_latitude,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount
0,2,2015-06-02 11:19:29,2015-06-02 11:47:52,1,1.63,-73.954430,40.764141,1,N,-73.974754,40.754093,2,17.0,0.0,0.5,0.00,0.0,0.3,17.80
1,2,2015-06-02 11:19:30,2015-06-02 11:27:56,1,0.46,-73.971443,40.758942,1,N,-73.978539,40.761909,1,6.5,0.0,0.5,1.00,0.0,0.3,8.30
2,2,2015-06-02 11:19:31,2015-06-02 11:30:30,1,0.87,-73.978111,40.738434,1,N,-73.990273,40.745438,1,8.0,0.0,0.5,2.20,0.0,0.3,11.00
3,2,2015-06-02 11:19:31,2015-06-02 11:39:02,1,2.13,-73.945892,40.773529,1,N,-73.971527,40.760330,1,13.5,0.0,0.5,2.86,0.0,0.3,17.16
4,1,2015-06-02 11:19:32,2015-06-02 11:32:49,1,1.40,-73.979088,40.776772,1,N,-73.982162,40.758999,2,9.5,0.0,0.5,0.00,0.0,0.3,10.30
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9994,1,2015-06-01 00:12:59,2015-06-01 00:24:18,1,2.70,-73.947792,40.814972,1,N,-73.973358,40.783638,2,11.0,0.5,0.5,0.00,0.0,0.3,12.30
9995,1,2015-06-01 00:12:59,2015-06-01 00:28:16,1,4.50,-74.004066,40.747818,1,N,-73.953758,40.779285,1,16.0,0.5,0.5,3.00,0.0,0.3,20.30
9996,2,2015-06-01 00:13:00,2015-06-01 00:37:25,1,5.59,-73.994377,40.766102,1,N,-73.903206,40.750546,2,21.0,0.5,0.5,0.00,0.0,0.3,22.30
9997,2,2015-06-01 00:13:02,2015-06-01 00:19:10,6,1.54,-73.978302,40.748531,1,N,-73.989166,40.762852,2,6.5,0.5,0.5,0.00,0.0,0.3,7.80


# Next up:

1. More practice with CSV
2. Other file formats
3. Retrieving from the Internet

# Exercise: Titanic

1. Load `titanic3.csv` into a data frame.
2. Find the mean age of people who survived.
3. Find the mean age of men who survived.
4. Find the mean age of women who survived, and who got on in S (Southhampton).